pkg/server/apiutil/rangeutil.go

```diff
@@ -0,0 +1,189 @@
+}
+
+// GetRangeIndexMappings translates a set of ordered ranges into a
+// RangeID -> []IndexNames mapping. It does this by executing the fololowing steps:
+//  1. Convert the set of ranges to a set of spans.
+//  2. Get the table descriptors that fall within the given spans.
+//  3. Get the database, table and index name for all indexes found in the descriptors.
+//  4. Return a mapping of the indexes which appear in each range.
+func GetRangeIndexMapping(
+	ctx context.Context,
+	txn descs.Txn,
+	codec keys.SQLCodec,
+	databases map[descpb.ID]catalog.DatabaseDescriptor,
+	ranges []roachpb.RangeDescriptor,
+) (map[roachpb.RangeID][]IndexNames, error) {
+	spans := RangesToTableSpans(codec, ranges)
+
+	tables, err := SpansToOrderedTableDescriptors(ctx, txn, codec, spans)
+	if err != nil {
+		return nil, err
+	}
+
+	indexes, err := TableDescriptorsToIndexNames(codec, databases, tables)
+	if err != nil {
+		return nil, err
+	}
+
+	return MapRangesToIndexes(ranges, indexes), nil
+}
+
+// MapRangesToIndexes is a utility function which iterates over two lists,
+// one consisting of ordered ranges, and the other consisting of ordered index names
+// and outputs a mapping from range to index.
+func MapRangesToIndexes(
+	ranges []roachpb.RangeDescriptor, indexes []IndexNames,
+) map[roachpb.RangeID][]IndexNames {
+	results := map[roachpb.RangeID][]IndexNames{}
+	contents := []IndexNames{}
+	flushToResults := func(rangeID roachpb.RangeID) {
+		results[rangeID] = contents
+		contents = []IndexNames{}
+	}
+
+	// move through the ranges + descriptors
+	// using two indexes, i, j.
+	// while i and j are valid
+	i := 0
+	j := 0
+	for i < len(ranges) && j < len(indexes) {
+		if spanFromRange(ranges[i]).Overlaps(indexes[j].Span) {
+			contents = append(contents, indexes[j])
+		}
+
+		if ranges[i].EndKey.AsRawKey().Less(indexes[j].Span.EndKey) {
+			flushToResults(ranges[i].RangeID)
+			i++
+		} else {
+			j++
+		}
+	}
+
+	if i < len(ranges) {
+		flushToResults(ranges[i].RangeID)
+	}
+	return results
+}
+
+// RangeToTableSpans is a simple utility function which converts a set of ranges
+// to a set of spans bound to the codec's SQL table space, and removed if the bound
+// span is zero length.
+func RangesToTableSpans(codec keys.SQLCodec, ranges []roachpb.RangeDescriptor) []roachpb.Span {
+	spans := []roachpb.Span{}
+
+	// cannot use keys.TableDataMin/Max
+	// Check the following: keys.TableDataMax.Less(keys.MakeSQLCodec(3).TablePrefix(1)) == true
+	bounds := roachpb.Span{
+		Key:    codec.TablePrefix(0),
+		EndKey: codec.TablePrefix(math.MaxUint32),
+	}
+	for _, rangeDesc := range ranges {
+		span := spanFromRange(rangeDesc).Clamp(bounds)
+		if !span.ZeroLength() {
+			spans = append(spans, span)
+		}
+	}
+
+	return spans
+}
+
+// SpansToOrderedTableDescriptors uses the transaction's collection to turn a set of
+// spans to a set of descriptors which describe the table space in which those spans lie.
+func SpansToOrderedTableDescriptors(
```

Comment:
Utility 2: A function which takes the above spans and uses the catalog
and new descriptor by span utility to turn those spans into a set of
table descriptors ordered by id (no tests for this func).